File: backend/core/imtihon.py
"""
DTM tayyorgarlik — natijalar serverda.

Savollar mijozda yasaladi (variant raqamidan, `frontend/src/lib/
imtihon.ts`); server faqat NATIJANI saqlaydi. Nega kerak: ilgari
tarix telefon xotirasida edi — ilovani o'chirgan yoki telefon
almashtirgan odam butun o'sish tarixini yo'qotardi, tayyorgarlikda esa
aynan shu tarix eng qimmatli narsa.

Ikki so'rov:

    yoz        bitta yoki bir nechta urinish (telefondagi eski tarix
               ham shu yo'l bilan bir marta ko'chiriladi)
    royxat     o'z natijalari: oxirgi urinishlar, har variantdagi eng
               yaxshisi va oxirgi beshtaning o'rtachasi
"""
from __future__ import annotations

from django.db import IntegrityError, transaction

from .models import ImtihonNatija, Profile

#: Variantlar soni — mijozdagi `VARIANTLAR` bilan bir xil.
VARIANTLAR = 12
#: Bitta variantda eng ko'p savol (ehtiyot chegarasi).
MAX_SAVOL = 60
#: Eng uzoq urinish — ikki soat (variant bir soatlik).
MAX_SEKUND = 2 * 3600
#: Bir so'rovda eng ko'p nechta urinish (eski tarixni ko'chirish uchun).
MAX_BIR_YOLA = 50
#: Ro'yxatda qaytariladigan oxirgi urinishlar.
OXIRGI = 20


def _butun(x, eng_kam: int, eng_kop: int) -> int | None:
    try:
        n = int(x)
    except (TypeError, ValueError):
        return None
    return n if eng_kam <= n <= eng_kop else None


def _toza(d: dict) -> dict | None:
    """Bitta urinishni tekshiradi. Yaroqsiz bo'lsa — `None` (jimgina tashlanadi)."""
    if not isinstance(d, dict):
        return None
    variant = _butun(d.get("variant"), 1, VARIANTLAR)
    jami = _butun(d.get("jami"), 1, MAX_SAVOL)
    if variant is None or jami is None:
        return None
    togri = _butun(d.get("togri"), 0, jami)
    vaqt = _butun(d.get("vaqt"), 1, 10 ** 14)
    if togri is None or vaqt is None:
        return None
    sekund = _butun(d.get("sekund"), 0, MAX_SEKUND)
    return {"variant": variant, "jami": jami, "togri": togri,
            "sekund": sekund if sekund is not None else 0, "mijoz_vaqt": vaqt}


def yoz(profil: Profile, urinishlar) -> int:
    """
    Urinishlarni yozadi. Nechta YANGI qator qo'shilganini qaytaradi.

    Takror kelgani (xuddi o'sha `vaqt`) jimgina tashlanadi — bu xato
    emas: telefon internet qaytganda qayta yuboradi.
    """
    if isinstance(urinishlar, dict):
        urinishlar = [urinishlar]
    if not isinstance(urinishlar, list):
        return 0
    yangi = 0
    for d in urinishlar[:MAX_BIR_YOLA]:
        t = _toza(d)
        if not t:
            continue
        try:
            with transaction.atomic():
                ImtihonNatija.objects.create(profile=profil, **t)
            yangi += 1
        except IntegrityError:
            continue
    return yangi


def _foiz(togri: int, jami: int) -> int:
    return round(100 * togri / jami) if jami else 0


def royxat(profil: Profile) -> dict:
    """O'z natijalari: oxirgilar, har variantdagi eng yaxshisi va o'rtacha."""
    qs = ImtihonNatija.objects.filter(profile=profil).order_by("-mijoz_vaqt")
    oxirgilar = [{
        "variant": n.variant, "togri": n.togri, "jami": n.jami,
        "sekund": n.sekund, "vaqt": n.mijoz_vaqt,
    } for n in qs[:OXIRGI]]

    eng: dict[int, dict] = {}
    for n in qs.only("variant", "togri", "jami"):
        bor = eng.get(n.variant)
        if not bor or n.togri > bor["togri"]:
            eng[n.variant] = {"togri": n.togri, "jami": n.jami}

    besh = oxirgilar[:5]
    ortacha = round(sum(_foiz(x["togri"], x["jami"]) for x in besh) / len(besh)) if besh else None
    return {
        "oxirgilar": oxirgilar,
        "eng_yaxshi": {str(k): v for k, v in eng.items()},
        "ortacha": ortacha,
        "jami": qs.count(),
    }

File: frontend/src/lib/imtihon.ts
/**
 * IMTIHON YO'LI — DTM matematika blokiga tayyorgarlik.
 *
 * ─────────────── NEGA ALOHIDA BO'LIM ───────────────
 *
 * Anketa shuni ko'rsatdi: kelganlarning 58% i talaba, yana 9–11
 * sinflar. Ular matematikani bitta aniq sabab bilan qidiradi —
 * IMTIHON. Ilovada esa hamma narsa SINF bo'yicha tizilgan: "11-sinf
 * matematika" degan kurs bor, lekin imtihonga tayyorgarlik yo'q edi.
 * Farqi kichik ko'rinadi, aslida katta:
 *
 *   Sinf kursi    bir yilning dasturi, bob-bob, tartib bilan.
 *   Imtihon       butun maktab kursidan aralash savol, vaqt bilan,
 *                 va eng muhimi — QAYSI MAVZUDA yiqilayotganingiz.
 *
 * ─────────────── VARIANT NIMA ───────────────
 *
 * Har variant — 30 savol, 60 daqiqa (haqiqiy imtihon sur'ati) va
 * 7–11 sinflarning hammasidan yig'iladi (`blok.IMTIHON_SINFLAR`).
 *
 * Variant RAQAMI bo'yicha yasaladi: 5-variant har qurilmada,
 * har safar aynan bir xil savollarni beradi. Sabab test to'plamidagi
 * bilan bir xil (`lib/toplam.ts`): natijalarni solishtirish uchun
 * hamma bitta testni ishlashi kerak. Savollar bazada saqlanmaydi —
 * faqat urug', qolganini generatorlar qiladi.
 */
import { bilanProfil, sorov } from "./api";
import { blokYasa } from "./blok";
import type { Blok } from "./blok";
import { kunUrugi, urugBilan } from "./oyin/urug";

/** Nechta variant bor. Ro'yxat uzun bo'lmasin — tanlov ham mehnat. */
export const VARIANTLAR = 12;

/** Bitta variantning o'lchami — DTM bilan bir xil. */
export const OLCHAM = { savol: 30, daqiqa: 60 };

/** Variant urug'i: raqam → o'zgarmas son. */
export const variantUrugi = (n: number): number => kunUrugi(`dtm-variant-${n}`, 7);

/**
 * Variantni yasaydi. Raqam noto'g'ri bo'lsa — `null`.
 *
 * `urugBilan` ichida yasaladi, ya'ni generatorlardagi `Math.random`
 * vaqtincha almashadi va natija takrorlanadi.
 */
export function variantYasa(n: number): Blok | null {
  if (!Number.isInteger(n) || n < 1 || n > VARIANTLAR) return null;
  return urugBilan(variantUrugi(n), () =>
    blokYasa(11, "dtm", { tur: "imtihon" }, OLCHAM));
}

/* ------------------------------------------------------------ natija */

/** Bitta topshirilgan variant — qurilmada saqlanadi. */
export interface ImtihonNatija {
  variant: number;
  togri: number;
  jami: number;
  sekund: number;
  vaqt: number;
}

const KALIT = "azapp_imtihon_v1";
const CHEK = 40;

export function natijalar(): ImtihonNatija[] {
  try {
    const xom = JSON.parse(localStorage.getItem(KALIT) || "[]") as ImtihonNatija[];
    return Array.isArray(xom) ? xom.filter((x) => x && typeof x.variant === "number") : [];
  } catch { return []; }
}

export function natijaSaqla(n: ImtihonNatija): void {
  try {
    const royxat = [n, ...natijalar()].slice(0, CHEK);
    localStorage.setItem(KALIT, JSON.stringify(royxat));
  } catch { /* xotira to'lgan — natija faqat ekranda qoladi */ }
}

/** Shu variantdagi ENG YAXSHI natija (yo'q bo'lsa — `null`). */
export function engYaxshi(variant: number): ImtihonNatija | null {
  const shu = natijalar().filter((x) => x.variant === variant);
  if (!shu.length) return null;
  return shu.reduce((a, b) => (b.togri > a.togri ? b : a));
}

/** Foiz — 0 dan 100 gacha. */
export const foiz = (n: Pick<ImtihonNatija, "togri" | "jami">): number =>
  (n.jami ? Math.round((100 * n.togri) / n.jami) : 0);

/**
 * O'RTACHA DARAJA — oxirgi beshta urinish bo'yicha.
 *
 * Bitta natija hech narsa demaydi: omad ham, charchoq ham bor.
 * Beshta urinishning o'rtachasi esa haqiqatga yaqin va aynan shu
 * son o'sib borishi kerak.
 */
export function daraja(): { foiz: number; urinish: number } | null {
  const oxirgi = natijalar().slice(0, 5);
  if (!oxirgi.length) return null;
  const jami = oxirgi.reduce((a, b) => a + foiz(b), 0);
  return { foiz: Math.round(jami / oxirgi.length), urinish: natijalar().length };
}

/* ------------------------------------------------------------ server */

/**
 * Serverdagi tarix — telefon xotirasidagi nusxadan USTUN.
 *
 * Nega: ilovani o'chirgan yoki telefon almashtirgan odam qurilmadagi
 * tarixni yo'qotadi, tayyorgarlikda esa aynan o'sish tarixi eng
 * qimmatli narsa. Qurilmadagi nusxa faqat internet yo'q paytda
 * ekranni bo'sh qoldirmaslik uchun turadi.
 */
export interface ServerTarix {
  oxirgilar: ImtihonNatija[];
  /** Variant raqami → eng yaxshi natija. */
  eng_yaxshi: Record<string, { togri: number; jami: number }>;
  /** Oxirgi beshtaning o'rtacha foizi (urinish bo'lmasa — `null`). */
  ortacha: number | null;
  jami: number;
}

/**
 * Qurilmadagi HAMMA urinishni serverga yuboradi va yakuniy tarixni
 * qaytaradi.
 *
 * Har ochilishda chaqiriladi va bu ATAYLAB: server takrorni o'zi
 * tashlaydi (`vaqt` bo'yicha), ya'ni eski tarix bir marta ko'chadi,
 * internetsiz ishlangan variant esa keyingi ochilishda yetib boradi.
 * Qurilmada ko'pi bilan 40 ta urinish turadi — so'rov yengil.
 */
export async function sinxronla(): Promise<ServerTarix> {
  // `bilanProfil` — oilaviy hisobda natija AYNAN tanlangan bolaga
  // yozilsin, hisobning birinchi profiliga emas.
  return sorov<ServerTarix>("/api/v1/imtihon/natija", bilanProfil({ urinishlar: natijalar() }));
}

/** Bitta tugagan urinishni yuboradi. Xato bo'lsa jim — keyingi `sinxronla` yetkazadi. */
export function serverga(n: ImtihonNatija): void {
  void sorov("/api/v1/imtihon/natija", bilanProfil({ ...n })).catch(() => {});
}
